scipy/special/_factorial.pxd

```diff
@@ -0,0 +1,8 @@
+from ._cephes cimport Gamma
+
+
+cdef inline double _factorial(double n) nogil:
```

Comment:
So here's a question: should this signature be `double _factorial(long double n)`? You could definitely get doubles through before, but I'm tempted to call that behavior broken. I'm all for pushing people to use `sc.gamma` if they actually want results for non-integral values.

Reply:
Hmm, I don't think that will lead to consistent behavior in `_factorial(n)` and then we'll need more type checking in `exact=True` branch, where `math.factorial(5.0)` will return `120` without error or warning.  

```
>>> import math
>>> type(math.factorial(5.0))
<class 'int'>
```
I'm not opposed to forcing input `n` to be `int` in `_factorial`, but if we do that I think it should be consistent behavior within the function. For example, currently, `math.factorial(5.1)` throws a `ValueError` but `math.factorial(5.0)` is the correct `int` returned. 
```
>>> math.factorial(5.1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: factorial() only accepts integral values
```
We'd either want to do something like in the legacy deprecations (e.g. the truncation in `bdtr` etc.) but here would be ***without*** a `RuntimeWarning` emitted. Or we could remove the `math.factorial(n)` call to get consistent behavior across the 2 branches of `exact`. Unless you see a third way that I'm not thinking about at the moment. 

Does that make sense? 

If so, I'm pretty much indifferent to the approach to get us there, so long as the two branches are consistent. 

If we're on the same page, which approach would you recommend?

Reply:
I looked at this today and is more clear how to proceed. So I'll do a check similar to:
https://github.com/scipy/scipy/blob/master/scipy/special/_legacy.pxd#L33-L37
but in a separate function and return a `ValueError` if `<int> x != x` I'll call the function something like `_int_arg_check()` and have the return message match that of the `math.factorial` this way we have consistent messaging on value errors across the `exact=True` and `exact=False` branches.

Reply:
Actually, what seems simpler is to put it directly into the `_factorial()` in cython, but it seems that the `ValueError` is ignored when thrown? I'm guessing this is some property of the build setup but I'm not clear if there is a way to override the behavior specifically for `_factorial()` or if it is better to place up in the python itself. @person142 can you give me a pointer on this? 

I pushed my changes if that helps to look at to understand. The message I get in the test is: 
```
===================================================================================== FAILURES ======================================================================================
__________________________________________________________________ TestFactorialFunctions.test_non_integral_inputs __________________________________________________________________
scipy/special/tests/test_basic.py:1842: in test_non_integral_inputs
    assert_raises(ValueError, special.factorial, 1.1, False)
E   Failed: DID NOT RAISE <class 'ValueError'>
------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------
ValueError: factorial() only accepts integral values
Exception ignored in: 'scipy.special._factorial._factorial'
ValueError: factorial() only accepts integral values
```

Reply:
Ok, so sorry to waffle here but yeah, I think your first instinct of "let's not change this" is correct, especially for this PR. The SciPy factorial function has been weird for a long time, so we shouldn't try and cram a fix for that in here.

Reply:
Ok, so to be clear, I'll revert the last commit and address the couple (minor) comments you have.